File: apps/agent/src/node-control.ts
import { execFile } from "node:child_process";
import { promisify } from "node:util";
import type { ControlAction } from "./control-actions.js";
import { sendFahControlCommand } from "./fah-control.js";
import { restartFoldopsAgent } from "./run-update.js";

const execFileAsync = promisify(execFile);

export interface ControlResult {
  ok: boolean;
  action: ControlAction;
  message: string;
  stdout: string;
  stderr: string;
}

export interface ControlStatus {
  foldops_agent: string;
  fah_client: string;
}

export async function getControlStatus(): Promise<ControlStatus> {
  return {
    foldops_agent: await systemdIsActive("foldops-agent"),
    fah_client: await systemdIsActive("fah-client"),
  };
}

async function systemdIsActive(unit: string): Promise<string> {
  try {
    const { stdout } = await execFileAsync("systemctl", ["is-active", unit]);
    return stdout.trim();
  } catch (err: unknown) {
    const execErr = err as { stdout?: string };
    return execErr.stdout?.trim() ?? "inactive";
  }
}

async function runSystemctl(
  args: ["start" | "stop" | "restart", string],
): Promise<{ stdout: string; stderr: string }> {
  const { stdout, stderr } = await execFileAsync("systemctl", args, {
    timeout: 60_000,
  });
  return { stdout: stdout.trim(), stderr: stderr.trim() };
}

export async function executeControlAction(
  action: ControlAction,
  opts: { allowReboot: boolean },
): Promise<ControlResult> {
  const fail = (
    message: string,
    stdout = "",
    stderr = "",
  ): ControlResult => ({
    ok: false,
    action,
    message,
    stdout,
    stderr,
  });

  try {
    switch (action) {
      case "agent.start": {
        const r = await runSystemctl(["start", "foldops-agent"]);
        return {
          ok: true,
          action,
          message: "foldops-agent started",
          stdout: r.stdout,
          stderr: r.stderr,
        };
      }
      case "agent.stop": {
        const r = await runSystemctl(["stop", "foldops-agent"]);
        return {
          ok: true,
          action,
          message: "foldops-agent stopped",
          stdout: r.stdout,
          stderr: r.stderr,
        };
      }
      case "agent.restart": {
        return {
          ok: true,
          action,
          message: "foldops-agent will restart",
          stdout: "",
          stderr: "",
        };
      }
      case "fah.start": {
        const r = await runSystemctl(["start", "fah-client"]);
        return {
          ok: true,
          action,
          message: "fah-client started",
          stdout: r.stdout,
          stderr: r.stderr,
        };
      }
      case "fah.stop": {
        const r = await runSystemctl(["stop", "fah-client"]);
        return {
          ok: true,
          action,
          message: "fah-client stopped",
          stdout: r.stdout,
          stderr: r.stderr,
        };
      }
      case "fah.restart": {
        const r = await runSystemctl(["restart", "fah-client"]);
        return {
          ok: true,
          action,
          message: "fah-client restarted",
          stdout: r.stdout,
          stderr: r.stderr,
        };
      }
      case "fah.pause": {
        const r = await sendFahControlCommand("pause");
        return {
          ok: r.ok,
          action,
          message: r.message,
          stdout: "",
          stderr: r.ok ? "" : r.message,
        };
      }
      case "fah.resume": {
        const r = await sendFahControlCommand("fold");
        return {
          ok: r.ok,
          action,
          message: r.ok ? "FAH folding resumed" : r.message,
          stdout: "",
          stderr: r.ok ? "" : r.message,
        };
      }
      case "fah.finish": {
        const r = await sendFahControlCommand("finish");
        return {
          ok: r.ok,
          action,
          message: r.ok ? "FAH finish command sent (completes WU then pauses)" : r.message,
          stdout: "",
          stderr: r.ok ? "" : r.message,
        };
      }
      case "host.reboot": {
        if (!opts.allowReboot) {
          return fail("Host reboot disabled (set CONTROLS_ALLOW_REBOOT=true)");
        }
        await execFileAsync("systemctl", ["reboot"], { timeout: 5_000 });
        return {
          ok: true,
          action,
          message: "Reboot initiated",
          stdout: "",
          stderr: "",
        };
      }
      default:
        return fail(`Unknown action: ${action}`);
    }
  } catch (err: unknown) {
    const execErr = err as {
      stdout?: string;
      stderr?: string;
      message?: string;
    };
    return fail(
      execErr.message ?? String(err),
      (execErr.stdout ?? "").trim(),
      (execErr.stderr ?? "").trim(),
    );
  }
}

/** Restart foldops-agent after HTTP response (agent.restart from HTTP handler). */
export function scheduleAgentSelfRestart(): void {
  setTimeout(() => {
    restartFoldopsAgent().catch((err) => {
      console.error("[control] agent restart failed:", err);
    });
  }, 400);
}

File: crates/foldops-agent/src/node_control.rs
use foldops_types::ControlAction;
use serde::Serialize;

use crate::fah::{send_fah_finish, send_fah_pause, send_fah_resume};

#[derive(Debug, Clone, Serialize)]
pub struct ControlResult {
    pub ok: bool,
    pub action: String,
    pub message: String,
    pub stdout: String,
    pub stderr: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct ControlStatus {
    pub foldops_agent: String,
    pub fah_client: String,
}

pub struct ControlContext {
    pub allow_reboot: bool,
    pub fah_ws_host: String,
    pub fah_ws_port: u16,
}

pub async fn get_control_status() -> ControlStatus {
    ControlStatus {
        foldops_agent: systemd_is_active("foldops-agent").await,
        fah_client: systemd_is_active("fah-client").await,
    }
}

async fn systemd_is_active(unit: &str) -> String {
    match tokio::process::Command::new("systemctl")
        .args(["is-active", unit])
        .output()
        .await
    {
        Ok(o) => String::from_utf8_lossy(&o.stdout).trim().to_string(),
        Err(_) => "inactive".into(),
    }
}

async fn run_systemctl(action: &str, unit: &str) -> Result<(String, String), String> {
    let output = tokio::process::Command::new("systemctl")
        .args([action, unit])
        .output()
        .await
        .map_err(|e| e.to_string())?;

    Ok((
        String::from_utf8_lossy(&output.stdout).trim().to_string(),
        String::from_utf8_lossy(&output.stderr).trim().to_string(),
    ))
}

pub async fn execute_control_action(
    action: ControlAction,
    ctx: &ControlContext,
) -> ControlResult {
    let action_str = action.as_str().to_string();

    let fail = |message: String, stdout: String, stderr: String| ControlResult {
        ok: false,
        action: action_str.clone(),
        message,
        stdout,
        stderr,
    };

    match action {
        ControlAction::AgentStart => match run_systemctl("start", "foldops-agent").await {
            Ok((stdout, stderr)) => ControlResult {
                ok: true,
                action: action_str,
                message: "foldops-agent started".into(),
                stdout,
                stderr,
            },
            Err(e) => fail(e, String::new(), String::new()),
        },
        ControlAction::AgentStop => match run_systemctl("stop", "foldops-agent").await {
            Ok((stdout, stderr)) => ControlResult {
                ok: true,
                action: action_str,
                message: "foldops-agent stopped".into(),
                stdout,
                stderr,
            },
            Err(e) => fail(e, String::new(), String::new()),
        },
        ControlAction::AgentRestart => ControlResult {
            ok: true,
            action: action_str,
            message: "foldops-agent will restart".into(),
            stdout: String::new(),
            stderr: String::new(),
        },
        ControlAction::FahStart => match run_systemctl("start", "fah-client").await {
            Ok((stdout, stderr)) => ControlResult {
                ok: true,
                action: action_str,
                message: "fah-client started".into(),
                stdout,
                stderr,
            },
            Err(e) => fail(e, String::new(), String::new()),
        },
        ControlAction::FahStop => match run_systemctl("stop", "fah-client").await {
            Ok((stdout, stderr)) => ControlResult {
                ok: true,
                action: action_str,
                message: "fah-client stopped".into(),
                stdout,
                stderr,
            },
            Err(e) => fail(e, String::new(), String::new()),
        },
        ControlAction::FahRestart => match run_systemctl("restart", "fah-client").await {
            Ok((stdout, stderr)) => ControlResult {
                ok: true,
                action: action_str,
                message: "fah-client restarted".into(),
                stdout,
                stderr,
            },
            Err(e) => fail(e, String::new(), String::new()),
        },
        ControlAction::FahPause => match send_fah_pause(&ctx.fah_ws_host, ctx.fah_ws_port).await {
            Ok(()) => ControlResult {
                ok: true,
                action: action_str,
                message: "FAH pause command sent".into(),
                stdout: String::new(),
                stderr: String::new(),
            },
            Err(msg) => fail(msg.clone(), String::new(), msg),
        },
        ControlAction::FahResume => match send_fah_resume(&ctx.fah_ws_host, ctx.fah_ws_port).await {
            Ok(()) => ControlResult {
                ok: true,
                action: action_str,
                message: "FAH folding resumed".into(),
                stdout: String::new(),
                stderr: String::new(),
            },
            Err(msg) => fail(msg.clone(), String::new(), msg),
        },
        ControlAction::FahFinish => match send_fah_finish(&ctx.fah_ws_host, ctx.fah_ws_port).await {
            Ok(()) => ControlResult {
                ok: true,
                action: action_str,
                message: "FAH finish command sent (completes WU then pauses)".into(),
                stdout: String::new(),
                stderr: String::new(),
            },
            Err(msg) => fail(msg.clone(), String::new(), msg),
        },
        ControlAction::HostReboot => {
            if !ctx.allow_reboot {
                return fail(
                    "Host reboot disabled (set CONTROLS_ALLOW_REBOOT=true)".into(),
                    String::new(),
                    String::new(),
                );
            }
            let _ = tokio::process::Command::new("systemctl")
                .arg("reboot")
                .output()
                .await;
            ControlResult {
                ok: true,
                action: action_str,
                message: "Reboot initiated".into(),
                stdout: String::new(),
                stderr: String::new(),
            }
        }
    }
}

pub fn schedule_agent_self_restart() {
    crate::update::schedule_agent_self_restart();
}
